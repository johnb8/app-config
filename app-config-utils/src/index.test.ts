import { isNode, isObject, isPrimitive, generateModuleText } from './index';

describe('isNode', () => {
  it('detects node.js env', () => {
    expect(isNode).toBe(true);
  });
});

describe('isObject', () => {
  it('marks an object as an object', () => {
    expect(isObject({})).toBe(true);
    expect(isObject([])).toBe(false);
    expect(isObject(null)).toBe(false);
    expect(isObject(42)).toBe(false);
    expect(isObject('foobar')).toBe(false);
  });
});

describe('isPrimitive', () => {
  it('marks primitives as such', () => {
    expect(isPrimitive(null)).toBe(true);
    expect(isPrimitive(42)).toBe(true);
    expect(isPrimitive('foobar')).toBe(true);
    expect(isPrimitive([])).toBe(false);
    expect(isPrimitive({})).toBe(false);
  });
});

describe('generateModuleText', () => {
  it('creates config module', () => {
    expect(
      generateModuleText(
        {
          foo: 'bar',
        },
        {
          environment: 'test',
          useGlobalNamespace: false,
          esmValidationCode: false,
          validationFunctionCode: undefined,
        },
      ),
    ).toMatchSnapshot();
  });

  it('creates config module with global namespace', () => {
    expect(
      generateModuleText(
        {
          foo: 'bar',
        },
        {
          environment: 'test',
          useGlobalNamespace: true,
          esmValidationCode: false,
          validationFunctionCode: undefined,
        },
      ),
    ).toMatchSnapshot();
  });

  it('creates config module with noBundledConfig', () => {
    expect(
      generateModuleText(undefined, {
        environment: 'test',
        useGlobalNamespace: true,
        esmValidationCode: false,
        validationFunctionCode: undefined,
      }),
    ).toMatchSnapshot();
  });

  it('creates config module with validation function', () => {
    expect(
      generateModuleText(
        {
          foo: 'bar',
        },
        {
          environment: 'test',
          useGlobalNamespace: true,
          esmValidationCode: false,
          // @ts-ignore
          validationFunctionCode: () => {
            return `
              const foo = 'bar';
            `;
          },
        },
      ),
    ).toMatchSnapshot();
  });

  it('creates config module with esm validation function', () => {
    expect(
      generateModuleText(
        {
          foo: 'bar',
        },
        {
          environment: 'test',
          useGlobalNamespace: true,
          esmValidationCode: true,
          // @ts-ignore
          validationFunctionCode: () => {
            return ['import foo from "bar";', `const foo = 'bar';`];
          },
        },
      ),
    ).toMatchSnapshot();
  });
});
