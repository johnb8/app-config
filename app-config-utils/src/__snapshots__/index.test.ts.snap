// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateModuleText creates config module 1`] = `
"
      const config = {\\"foo\\":\\"bar\\"};

      export { config };
      export default config;
    
      export function currentEnvironment() {
        return \\"test\\";
      }
    "
`;

exports[`generateModuleText creates config module with esm validation function 1`] = `
"
      const globalNamespace = (typeof window === 'undefined' ? globalThis : window) || {};

      const configValue = {\\"foo\\":\\"bar\\"};

      // if the global was already defined, use it
      const config = (globalNamespace._appConfig || configValue);

      // if the global is frozen then it was set by electron and we can't change it, but we'll set it if we can
      if (
        typeof globalNamespace._appConfig === 'undefined' ||
        !Object.isFrozen(globalNamespace._appConfig)
      ) {
        globalNamespace._appConfig = config;
      }

      export { config };
      export default config;
    
        const foo = 'bar';

        
        function genValidateConfig(){
          const validateConfigModule = {};
          (function(module){import foo from \\"bar\\";})(validateConfigModule);
          return validateConfigModule.exports;
        }

        
        export const validateConfig = /*#__PURE__*/ genValidateConfig();
      
      export function currentEnvironment() {
        return globalNamespace._appConfigEnvironment || \\"test\\";
      }
    "
`;

exports[`generateModuleText creates config module with global namespace 1`] = `
"
      const globalNamespace = (typeof window === 'undefined' ? globalThis : window) || {};

      const configValue = {\\"foo\\":\\"bar\\"};

      // if the global was already defined, use it
      const config = (globalNamespace._appConfig || configValue);

      // if the global is frozen then it was set by electron and we can't change it, but we'll set it if we can
      if (
        typeof globalNamespace._appConfig === 'undefined' ||
        !Object.isFrozen(globalNamespace._appConfig)
      ) {
        globalNamespace._appConfig = config;
      }

      export { config };
      export default config;
    
      export function currentEnvironment() {
        return globalNamespace._appConfigEnvironment || \\"test\\";
      }
    "
`;

exports[`generateModuleText creates config module with noBundledConfig 1`] = `
"
      const globalNamespace = (typeof window === 'undefined' ? globalThis : window) || {};

      const config = globalNamespace._appConfig;

      if (typeof config === 'undefined') {
        throw new Error('Config is not loaded in _appConfig');
      }

      export { config };
      export default config;
    
      export function currentEnvironment() {
        return globalNamespace._appConfigEnvironment || \\"test\\";
      }
    "
`;

exports[`generateModuleText creates config module with validation function 1`] = `
"
      const globalNamespace = (typeof window === 'undefined' ? globalThis : window) || {};

      const configValue = {\\"foo\\":\\"bar\\"};

      // if the global was already defined, use it
      const config = (globalNamespace._appConfig || configValue);

      // if the global is frozen then it was set by electron and we can't change it, but we'll set it if we can
      if (
        typeof globalNamespace._appConfig === 'undefined' ||
        !Object.isFrozen(globalNamespace._appConfig)
      ) {
        globalNamespace._appConfig = config;
      }

      export { config };
      export default config;
    
        
        function genValidateConfig(){
          const validateConfigModule = {};
          (function(module){
              const foo = 'bar';
            })(validateConfigModule);
          return validateConfigModule.exports;
        }

        
        export const validateConfig = /*#__PURE__*/ genValidateConfig();
      
      export function currentEnvironment() {
        return globalNamespace._appConfigEnvironment || \\"test\\";
      }
    "
`;
